modules/project_cleanup/function_source/main.go

```diff
@@ -17,99 +17,237 @@ limitations under the License.
+	labels := make(map[string]string)
+	err := json.Unmarshal([]byte(targetExcludedLabels), &labels)
 	if err != nil {
-		return err
+		logger.Printf("Fail to get labels map from [%s] env variable, error [%s]", envVariableName, err.Error())
+	} else {
+		logger.Printf("Got labels map [%s] from [%s] env variable", labels, envVariableName)
+	}
+	return labels
+}
+
+func getCorrectFolderIdOrTerminateExecution() string {
+	targetFolderIdString := os.Getenv(TargetFolderId)
+	matched, err := regexp.MatchString(targetFolderRegexp, targetFolderIdString)
+	if err != nil || !matched {
+		logger.Fatal(fmt.Sprintf("Invalid folder id [%s]. Specify correct value, Please.", targetFolderIdString))
 	}
+	return targetFolderIdString
+}
 
-	cloudResourceManagerService, err := cloudresourcemanager.New(c)
+func getResourceManagerServiceOrTerminateExecution(ctx context.Context) *cloudresourcemanager.Service {
+	logger.Println("Try to initialize Google client")
+	client, err := google.DefaultClient(ctx, cloudresourcemanager.CloudPlatformScope)
 	if err != nil {
-		return err
+		logger.Println("Fail to initialize Google client, terminate execution")
+		logger.Fatal(err)
 	}
+	logger.Println("Initialized Google client")
+	logger.Println("Try to get Cloud Resource Manager")
+	cloudResourceManagerService, err := cloudresourcemanager.New(client)
+	if err != nil {
+		logger.Println("Fail to get Cloud Resource Manager, terminate execution")
+		logger.Fatal(err)
+	}
+	logger.Println("Got Cloud Resource Manager")
+	return cloudResourceManagerService
+}
 
-	resourceCreationCutoff := tooOldTime(int64(acceptableAgeInHours * 60 * 60))
+func getFolderServiceOrTerminateExecution(ctx context.Context) *cloudresourcemanager2.FoldersService {
+	logger.Println("Try to initialize Google client")
+	client, err := google.DefaultClient(ctx, cloudresourcemanager.CloudPlatformScope)
+	if err != nil {
+		logger.Println("Fail to initialize Google client, terminate execution")
+		logger.Fatal(err)
+	}
+	logger.Println("Initialized Google client")
+	logger.Println("Try to get Folders Service")
+	cloudResourceManagerService, err := cloudresourcemanager2.New(client)
+	if err != nil {
+		logger.Println("Fail to get Folders Service, terminate execution")
+		logger.Fatal(err)
+	}
+	logger.Println("Got Folders Service")
+	return cloudResourceManagerService.Folders
+}
 
-	logger.Println("Looking through projects")
-	req := cloudResourceManagerService.Projects.List()
-	if err := req.Pages(ctx, func(page *cloudresourcemanager.ListProjectsResponse) error {
-		for _, project := range page.Projects {
-			if val, ok := project.Labels[key]; !(ok && val == value) {
-				logger.Println(fmt.Sprintf("Rejecting project %s because of missing label", project.ProjectId))
-				continue
-			}
+func invoke(ctx context.Context) {
+	cloudResourceManagerService := getResourceManagerServiceOrTerminateExecution(ctx)
+	folderService := getFolderServiceOrTerminateExecution(ctx)
+
+	removeLien := func(name string) {
```

Comment:
what is lien?

Reply:
https://cloud.google.com/resource-manager/docs/project-liens